src/server.rs

```diff
@@ -402,3 +403,41 @@ pub extern "C" fn rustls_server_session_write_tls(
         rustls_result::Ok
     }
 }
+
+/// Write up to `len` characters of the SNI hostname into `buf`. If the
+/// handshake has not been performed yet or if the client does not support SNI,
+/// the returned length in out_n will be 0.
+/// https://docs.rs/rustls/0.19.0/rustls/struct.ServerSession.html#method.get_sni_hostname
+#[no_mangle]
+pub extern "C" fn rustls_server_session_sni_hostname_get(
+    session: *mut rustls_server_session,
+    buf: *mut c_char,
+    len: size_t,
+    out_n: *mut size_t,
+) -> rustls_result {
+    ffi_panic_boundary! {
+        let session: &mut ServerSession = try_ref_from_ptr!(session, &mut ServerSession);
+        let write_buf: &mut [u8] = unsafe {
+            let out_n: &mut size_t = match out_n.as_mut() {
+                Some(out_n) => out_n,
+                None => return NullParameter,
+            };
+            *out_n = 0;
+            if buf.is_null() {
+                return NullParameter;
+            }
+            slice::from_raw_parts_mut(buf as *mut u8, len as usize)
+        };
+        let hostname_str = match session.get_sni_hostname() {
+            Some(hostname_str) => hostname_str,
+            None => "",
+        };
+        let len: usize = min(write_buf.len() - 1, hostname_str.len());
+        write_buf[..len].copy_from_slice(&hostname_str.as_bytes()[..len]);
```

Comment:
Would this truncate the `hostname` if the `buf` provided by the C side is too small? Is this the contract between `curl` and TLS backends? Silently truncating the hostname seems risky to me. Can we instead return an error indicating to the caller explicitly that they need to provide a bigger buffer?

Reply:
This would truncate, but the new way @jsha proposed would get rid of that buffer anyway.